### DISCUSSION TITLE
Parsing and visualizing diff output

### DISCUSSION THREAD

**Discussion by simoneruffini** [neovim/neovim#27154] [Ideas]
vim/neovim can be used as a diff viewer, but that configuration works per file and must be configured.
It would be awesome to simply pipe the output of a diff command to neovim and be able to show the direct result. My usecase is in diffing directories: where vimdiff would just return a per file diff a custon parser of the diff output  will show a continous stream of text of the diffs. This feature is provided by tools like [delta](https://github.com/dandavison/delta) or [meld](http://meldmerge.org/). What do you think? In my opinion a feature like this should simplify the steps necessary to configure vim as a diff viewer because you could just type `: readiff !diff -r dira dirb` where `readiff` is the internal function to do the parsing

**Comment by lewis6991** [comment]
Sounds like a potential plugin, not a core feature.

**Reply by simoneruffini** [reply]
I don't agree, neovim already has the function to diff files and you can use neovim as a diff file viewer as the output of git-diff but the process to do that is convoluted (in my opinion) and with the limitation of per file diff (you need to diff one file at a time and can't go back and forth quickly between diffs (in my experience)). So to me it makes perfectly sense to add this feature, it would make the diffing process much easier (for example `:neovimdiff !git diff %` or `:neovimdiff !diff folderA folderB`) and appealing to other users that simply need to view a diff from the diff command (that is not very readable by itself). Furthermore is more handy to have a unique view of the diff instead a per file diff (output of delta in the previous post).  As of now you need other tools/software to examine diffs: emacs seems to implement this thing natively (but rough around the edges to get better performance you still need a plugin)

**Reply by dundargoc** [reply]
Tbf, having it as a plugin first significantly increases its chances of it getting added in core since the implementation will already be done and its usefulness can be directly demonstrated. It's still not guaranteed to be included obviously, but I wouldn't count it out.

**Reply by lewis6991** [reply]
@simoneruffini nothing you said provides a reason why this should be built in. All you've done is explain further on why you would like it. Every new feature incurs both an initial implementation cost and an ongoing maintenance cost. Just because somebody thinks some elaborate idea would be a "nice to have" just isn't a strong enough justification.

If this is something you would like, then Neovim already provides all the tools to implement this as a plugin, then you can maintain it and address all the issues and feature requests people will raise.

**Reply by simoneruffini** [reply]
@lewis6991 What do you mean with reason? The reason is simple: provide functionality to the user in particular a use-case that in my opinion is frequently used i.e. developers frequently diff stuff. And my point is that the author of vim already thought diff was an important tool for file manipualtion: the functionality is already there! but it did not extend it to its full potential.

**Reply by lewis6991** [reply]
Adding every suggestion to "provide functionality to the user in particular a use-case" **does not scale**. This is **exactly** why we have a plugin eco-system and why one of the explicit goals of Neovim is extensibility.

**Reply by simoneruffini** [reply]
One of the best features of vim is that is full of small functionalities, ex. the convert diff to html.  For you this is a totally does-not-scale feature that is still present in neovim. IMO you can't apply the does-not-scale as you do otherwise you provide nothing and leave everything to the user to implement.
EDIT:
moreover I'm not talking about adding an entirely new feature, but leveraging what already exists... diff files are not a particular use-case, diffing as it-is is a first class citizen of neovim/vim (used in git merges for example) so I don't understand why to use that feature you need to be in a particular context (for example git merges) and can't be used independently. Neovim/vim has the power to understand diffs!

**Reply by clason** [reply]
Proof of concept (as a plugin, unless you have a very clear and convincing explanation why it _needs_ -- not "can", not "should", but **needs** -- to be implemented in C) will go a long way towards showing feasibility. Right now, you want something but want others to pay the cost for it. As none of the maintainers want this, this means it won't get done -- this is just how open source works. And no wall of text will change that.

**Reply by simoneruffini** [reply]
Ok, yes I'm ok with that!
